16_game.js

```diff
@@ -0,0 +1,636 @@
+..................................................############################################################
+..............................................................................................................
+`,
+];
+
+var Level = class Level {
+    constructor(plan) {
+        let rows = plan
+            .trim()
+            .split('\n')
+            .map(l => [...l]);
+        this.height = rows.length;
+        this.width = rows[0].length;
+        this.startActors = [];
+
+        this.rows = rows.map((row, y) => {
+            return row.map((ch, x) => {
+                let type = levelChars[ch];
+                if (typeof type == 'string') return type;
+                this.startActors.push(type.create(new Vec(x, y), ch));
+                return 'empty';
+            });
+        });
+    }
+};
+
+var State = class State {
+    constructor(level, actors, status) {
+        this.level = level;
+        this.actors = actors;
+        this.status = status;
+    }
+
+    static start(level) {
+        return new State(level, level.startActors, 'playing');
+    }
+
+    get player() {
+        return this.actors.find(a => a.type == 'player');
+    }
+};
+
+var Vec = class Vec {
+    constructor(x, y) {
+        this.x = x;
+        this.y = y;
+    }
+    plus(other) {
+        return new Vec(this.x + other.x, this.y + other.y);
+    }
+    times(factor) {
+        return new Vec(this.x * factor, this.y * factor);
+    }
+};
+
+var Player = class Player {
+    constructor(pos, speed) {
+        this.pos = pos;
+        this.speed = speed;
+    }
+
+    get type() {
+        return 'player';
+    }
+
+    static create(pos) {
+        return new Player(pos.plus(new Vec(0, -0.5)), new Vec(0, 0));
+    }
+};
+
+Player.prototype.size = new Vec(0.8, 1.5);
+
+//My solution - Monster
+class Monster {
+    constructor(pos, speed) {
+        this.pos = pos;
+        this.speed = speed;
+    }
+
+    get type() {
+        return 'monster';
+    }
+
+    static create(pos) {
+        return new Monster(pos.plus(new Vec(0, 0.2)), 1);
+    }
+
+    update(time, state) {
+        const moveMonster = this.pos.plus(new Vec(this.speed * time, 0));
+        if (state.level.touches(moveMonster, this.size, 'wall'))
+            return new Monster(moveMonster, -this.speed);
+        return new Monster(moveMonster, this.speed);
+    }
+
+    collide(state) {
+        const playerY = state.player.pos.y + state.player.size.y;
+        const monsterY = this.pos.y + this.size.y;
+        if (playerY < monsterY) {
+            const index = state.actors.indexOf(this);
+            state.actors.splice(index, 1);
+            return new State(state.level, state.actors, state.status);
+        }
+        return new State(state.level, state.actors, 'lost');
+    }
+}
+
+Monster.prototype.size = new Vec(1.2, 0.8);
+//---------------------------------------------------------
+
+var Lava = class Lava {
```

Comment:
is this your code?

Reply:
Line 295 is the last line of monster, which is my code.  Below (line 570) there are two more solutions.